# Creating a flag that sorts products

In [1]:
import pandas as pd
import os

path = '/Users/mariatirado/29-10-2025 Instacart Basket Analysis/02 Data/Prepared Data'

ords_prods_merge = pd.read_parquet(os.path.join(path, 'ords_prods_merge.parquet'))

print(ords_prods_merge.shape)
ords_prods_merge.head()

(32432460, 14)


,order_id,user_id,eval_set,order_number,order_dow,order_hour_of_day,days_since_prior_order,product_id,add_to_cart_order,reordered,product_name,aisle_id,department_id,prices
0,2539329,1,prior,1,2,8,NaN,196,1,0,Soda,77,7,9.0
1,2539329,1,prior,1,2,8,NaN,14084,2,0,Organic Unsweetened Vanilla Almond Milk,91,16,12.5
2,2539329,1,prior,1,2,8,NaN,12427,3,0,Original Beef Jerky,23,19,4.4
3,2539329,1,prior,1,2,8,NaN,26088,4,0,Aged White Cheddar Popcorn,23,19,4.7
4,2539329,1,prior,1,2,8,NaN,26405,5,0,XL Pick-A-Size Paper Towel Rolls,54,17,1.0


In [2]:
df = ords_prods_merge[:1000000]
print(df.shape)

(1000000, 14)


In [3]:
# Defining function 

def price_label(row):

  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High range product'
  else: return 'Not enough data'

In [4]:
df['price_range'] = df.apply(price_label, axis=1)

/var/folders/zj/m82l3gps79nc61v2flg2kkzr0000gn/T/ipykernel_7023/2848743372.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1)


In [5]:
df['price_range'].value_counts(dropna = False)

price_range
Mid-range product     673470
Low-range product     314117
High range product     12413
Name: count, dtype: int64

# If-Statements with the loc() Function

In [7]:
df = df.copy()

In [8]:
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [9]:
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product' 

In [10]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [12]:
df.loc[:, 'price_range_loc'].value_counts(dropna=False)

price_range_loc
Mid-range product     673470
Low-range product     314117
High-range product     12413
Name: count, dtype: int64

# If-Statement with loc() for ords_prods_merge

In [13]:
ords_prods_merge.loc[ords_prods_merge['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [14]:
ords_prods_merge.loc[
    (ords_prods_merge['prices'] <= 15) & (ords_prods_merge['prices'] > 5),
    'price_range_loc'
] = 'Mid-range product'

In [15]:
ords_prods_merge.loc[ords_prods_merge['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [16]:
ords_prods_merge['price_range_loc'].value_counts(dropna=False)

price_range_loc
Mid-range product     21889001
Low-range product     10125777
High-range product      417682
Name: count, dtype: int64

# If-Statements with For-Loops

In [18]:
ords_prods_merge['order_dow'].value_counts(dropna = False)

order_dow
0    6209268
1    5665509
6    4500014
2    4217540
5    4209264
3    3843871
4    3786994
Name: count, dtype: int64

In [21]:
result = []

for value in ords_prods_merge['order_dow']:
  if value == 0:
    result.append('Busiest day')
  elif value == 4:
    result.append('Least busy')
  else:
    result.append('Regularly busy')

In [22]:
result

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy

In [23]:
ords_prods_merge['day_type'] = result

In [24]:
ords_prods_merge['day_type'].value_counts(dropna=False)

day_type
Regularly busy    22436198
Busiest day        6209268
Least busy         3786994
Name: count, dtype: int64

# “Busiest day” to  “Busiest days” (plural)

In [25]:
ords_prods_merge['order_dow'].value_counts()

order_dow
0    6209268
1    5665509
6    4500014
2    4217540
5    4209264
3    3843871
4    3786994
Name: count, dtype: int64

In [27]:
import numpy as np

ords_prods_merge['busiest_days_new'] = np.where(
    ords_prods_merge['order_dow'].isin([0, 1]), 'Busiest days',  # top 2
    np.where(
        ords_prods_merge['order_dow'].isin([3, 4]), 'Slowest days',  # bottom 2
        'Regularly busy'))

In [28]:
ords_prods_merge['busiest_days_new'].value_counts(dropna=False)

busiest_days_new
Regularly busy    12926818
Busiest days      11874777
Slowest days       7630865
Name: count, dtype: int64

# Answer
According to the results:
Busiest days (0 and 1) → 11,874,777 orders, showing these are the two most active shopping days.

Slowest days (3 and 4) → 7,630,865 orders, confirming they have the least shopping activity.

Regularly busy days → 12,926,818 orders, representing the remaining days with steady traffic.

The difference between the busiest and slowest days suggests strong weekly purchase patterns — likely linked to weekend and start-of-week shopping habits.

These insights can help optimize promotions and staff scheduling, focusing resources on the high-demand days.

# Busiest hours of the day

In [29]:
ords_prods_merge['order_hour_of_day'].value_counts(dropna=False)

order_hour_of_day
10    2764238
11    2738418
14    2691398
15    2664370
13    2663115
12    2620668
16    2537310
9     2456553
17    2089340
8     1719870
18    1637825
19    1259320
20     976980
7      891886
21     796332
22     634706
23     402592
6      290762
0      218922
1      115779
5       88052
2       69428
4       53279
3       51317
Name: count, dtype: int64

In [30]:
import numpy as np

ords_prods_merge['busiest_period_of_day'] = np.where(
    (ords_prods_merge['order_hour_of_day'] >= 17) & (ords_prods_merge['order_hour_of_day'] <= 23), 'Most orders',
    np.where(
        (ords_prods_merge['order_hour_of_day'] >= 7) & (ords_prods_merge['order_hour_of_day'] <= 16), 'Average orders',
        'Fewest orders'))

In [31]:
ords_prods_merge['busiest_period_of_day'].value_counts(dropna=False)

busiest_period_of_day
Average orders    23747826
Most orders        7797095
Fewest orders       887539
Name: count, dtype: int64

In [32]:
ords_prods_merge.groupby('busiest_period_of_day')['order_hour_of_day'].unique()

busiest_period_of_day
Average orders    [8, 7, 12, 15, 9, 14, 16, 11, 10, 13]
Fewest orders                     [0, 5, 4, 6, 1, 2, 3]
Most orders                [19, 18, 17, 20, 21, 22, 23]
Name: order_hour_of_day, dtype: object

# Answer
The “Most orders” period occurs between 19:00 and 23:00, confirming that evenings are the busiest hours for Instacart users.

The “Average orders” period covers 07:00–16:00, showing steady activity throughout daytime hours.

The “Fewest orders” period happens between 00:00 and 06:00, when demand is significantly lower.

Most Instacart orders are placed during the Average orders window (≈23.7M orders), followed by the Most orders period (≈7.8M), and very few during Fewest orders (≈0.9M).

These insights can help optimize server capacity and staffing, reducing the risk of app slowdowns during the evening peaks.

# Export

In [33]:
import os

path_pickle = r'/Users/mariatirado/29-10-2025 Instacart Basket Analysis/02 Data/Prepared Data'

pickle_path = os.path.join(path_pickle, 'ords_prods_merge_updated.pkl')


ords_prods_merge.to_pickle(pickle_path)

print(f"✅ Pickle file saved at: {pickle_path}")

✅ Pickle file saved at: /Users/mariatirado/29-10-2025 Instacart Basket Analysis/02 Data/Prepared Data/ords_prods_merge_updated.pkl
